static/js/containers/pages/CheckoutPage.js

```diff
@@ -58,31 +59,45 @@ export const calcSelectedRunIds = (
   }
 
   const selectedRunIds = {}
-  for (const course of item.courses) {
-    if (course.next_run_id) {
-      selectedRunIds[course.id] = course.next_run_id
-    }
-  }
-
-  const courseLookup = {}
-  for (const course of item.courses) {
-    for (const run of course.courseruns) {
-      courseLookup[run.id] = course.id
-    }
-  }
 
-  // Try to preselect a run if the ID was given
   if (preselectId) {
-    const courseId = courseLookup[preselectId]
-    selectedRunIds[courseId] = preselectId
-  }
+    let preselectCourseId, preselectedRuneDate
 
-  for (const runId of item.run_ids) {
-    const courseId = courseLookup[runId]
+    const courseLookup = {}
+    for (const course of item.courses) {
+      courseLookup[course.id] = {}
+      for (const run of course.courseruns) {
+        if (run.id === preselectId) {
+          preselectCourseId = course.id
+          preselectedRuneDate = run.start_date
+          selectedRunIds[course.id] = run.id
+        }
+        courseLookup[course.id][run.id] = run.start_date
+      }
+    }
 
-    // there should only be one run selected for a course
-    selectedRunIds[courseId] = runId
+    for (const [courseId, courseRuns] of Object.entries(courseLookup)) {
+      if (courseId !== String(preselectCourseId)) {
+        for (const [runId, runDate] of Object.entries(courseRuns)) {
+          if (
+            moment(runDate)
+              .startOf("day")
+              .isSameOrAfter(moment(preselectedRuneDate).startOf("day"))
+          ) {
+            selectedRunIds[courseId] = runId
+            break
+          }
+        }
+      }
+    }
+  } else {
+    for (const course of item.courses) {
+      if (course.next_run_id) {
+        selectedRunIds[course.id] = course.next_run_id
+      }
+    }
```

Comment:
There's a lot of nesting here and it could be more readable. This code snippet below could be optimized a little, but I would suggest something like this:

```javascript
// static/js/lib/util.js
export const sameDayOrLater = (momentDate1: Moment, momentDate2: Moment): boolean => (
	momentDate1.startOf("day").isSameOrAfter(momentDate2.startOf("day"))
)


// static/js/containers/pages/CheckoutPage.js
let selectedRunIds = {}
let preselectCourseId, preselectedRunId, preselectedRunDate
if (preselectId) {
  for (const course of item.courses) {
    let matchingRun = course.courseruns.find(
      (run) => run.id === preselectId
    )
    if (matchingRun) {
      preselectCourseId = course.id
      preselectedRunDate = moment(matchingRun.start_date)
      selectedRunIds[preselectCourseId] = matchingRun.id
      break
    }
  }
}
if (!preselectCourseId) {
	for (const course of item.courses) {
    if (course.next_run_id) {
      selectedRunIds[course.id] = course.next_run_id
    }
  }
  return selectedRunIds
}

const otherCourses = R.reject(R.propEq("id", preselectCourseId), item.courses)
for (const course of otherCourses) { 
  let runClosestToPreselect = course.courseruns.find(
    (run) => sameDayOrLater(moment(run.start_date), preselectedRunDate)
  )
  let selectedRunId = runClosestToPreselect ? runClosestToPreselect.id : course.next_run_id
  if (selectedRunId) {
    selectedRunIds[course.id] = selectedRunId
  }
}
return selectedRunIds

```